webpack/webpack5-localization-plugin/src/TrueHashPlugin.ts

```diff
@@ -0,0 +1,401 @@
+                );
+              } else {
+                let assetSource: sources.Source = asset.source;
+                const assetName: string = asset.name;
+                if (chunkDependencies!.size > 0) {
+                  const relevantHashReplacements: Map<string, string> = new Map();
+                  let hasAnyReplacements: boolean = false;
+                  let allReplacementsAreTheSameLengthAsOriginals: boolean = true;
+                  for (const dependency of chunkDependencies!) {
+                    const asyncChunkHashReplacements: IHashReplacement | undefined =
+                      hashReplacementsByChunk.get(dependency);
+                    if (!asyncChunkHashReplacements) {
+                      compilation.errors.push(
+                        new thisWebpack.WebpackError(
+                          `Could not find hash replacements for chunk ${dependency.id}.`
+                        )
+                      );
+                    } else {
+                      const { existingHash: otherChunkExistingHash, trueHashByLocale } =
+                        asyncChunkHashReplacements;
+                      let replacementHash: string | undefined;
+                      if (typeof trueHashByLocale === 'object') {
+                        if (locale) {
+                          replacementHash = trueHashByLocale[locale];
+                        }
+                      } else {
+                        replacementHash = trueHashByLocale;
+                      }
+
+                      if (replacementHash) {
+                        if (relevantHashReplacements.has(otherChunkExistingHash)) {
+                          compilation.errors.push(
+                            new thisWebpack.WebpackError(
+                              `Found multiple replacements for hash ${otherChunkExistingHash} ` +
+                                `in chunk ${chunk.id}.`
+                            )
+                          );
+                        } else {
+                          allReplacementsAreTheSameLengthAsOriginals &&=
+                            replacementHash.length === otherChunkExistingHash.length;
+                          relevantHashReplacements.set(otherChunkExistingHash, replacementHash);
+                          hasAnyReplacements = true;
+                        }
+                      }
+                    }
+                  }
+
+                  if (hasAnyReplacements) {
+                    const sourceString: string = assetSource.source().toString();
+                    const replaceSource: sources.ReplaceSource = new thisWebpack.sources.ReplaceSource(
+                      assetSource,
+                      assetName
+                    );
+
+                    if (allReplacementsAreTheSameLengthAsOriginals) {
+                      // If all of the replacements are the same length as the originals, we can walk the string
```

Comment:
`ReplaceSource` takes care of ordering the replacements so that the indices are correct. The call to `.replace` just queues the replacement based on indices in the original source.